### ISSUE TITLE
Migrating from TextField/CharField to RichTextField should insert paragraph tags around text

### ISSUE THREAD

**Issue by alexgleason** [torchbox/wagtail#2290]
Just ran into a problem where the client populated ~50 pages of content then requested that a CharField be changed into a RichTextField.

But in the database, none of the existing entries for that field have <p> tags around them. Yet, when I revisit the editor page and click "submit," the <p> tags will be added, since that's what the editor does.

So I think during the migration, it should wrap these fields in <p> tags.

(Ideally, it could also do the same thing going the other way. But that's a different can of worms.)

**Comment by timheap** [comment]
I dont think this is possible to do automatically. Django does not seem to have any inbuilt support for automatically migrating the data in a field. However, you can make your migration do this as part of the migration quite easily. Something like the following would work:

```python
from django.db import migrations
from django.db.models import F

def wrap_in_paragraph(apps, schema_editor):
    if not schema_editor.connection.alias == 'default':
        return
    ContentPage = apps.get_model('content', 'ContentPage')
    ContentPage.objects.update(body=('<p>' + F('body') + '</p>'))

class Migration(migrations.Migration):
    dependencies = [
        # Dependencies to other migrations
    ]
    operations = [
        migrations.RunPython(wrap_in_paragraph),
    ]
```

This can be merged with the migration that converts the CharField to a RichTextField by appending the `migrations.RunPython(wrap_in_paragraph)` to the `operations` list of that migration.

**Comment by thibaudcolas** [comment]
I initially marked this as "related" when reviewing rich text issues a few weeks ago, but there won't be any changes to this because of the new editor as the storage format stays the same.

Should this workaround be added into the docs?

**Comment by thibaudcolas** [comment]
For future reference – based on my work in #8930, I believe this migration is (a bit) more complex than suggested above. For CharField, we need to add the wrapping `<p>` and escape all HTML typed inside the content. For TextField, we need to wrap _each line_ in a paragraph (`\n\n`), and also convert soft line breaks (single `\n`) into `<br>` tags.

We’ve also recently introduced character counts for rich text that discard markup but don’t have the same counting logic for line breaks as plain-text fields. With equivalent content, I think a rich text field would count its length shorter than a TextField, so we should be good, but it’d be worth checking with edge cases in mind.

This is a complex enough operation that to me it would warrant being supported in core even if it’s not done automatically – just document the approach and provide the right method for this specific conversion.

See also https://github.com/wagtail/wagtail/discussions/8156 / https://github.com/sandilsranasinghe/wagtail-streamfield-migration-toolkit, which also explores migration utilities (though for StreamField, not rich text).

**Comment by mixxorz** [comment]
Here's a simple snippet to convert plain text to richtext:

```python
def convert_plain_text_to_rich_text(text):
    """Convert plain text to rich text"""
    result = "<p>"

    prev_line = ""
    for line in text.splitlines():
        if line:
            if prev_line:
                result += "<br/>"

            result += line
        else:
            if prev_line:
                result += "</p><p>"
            else:
                result += "<br/>"

        prev_line = line

    result += "</p>"

    return result
```

Example input:

```
Prevent prevent here box speak level.
Like send serve remember beyond newspaper effort debate.

Instead choose option contain air less various.
Happen body nearly kitchen growth purpose those like.


Involve still along stock series spend.
```

Example output:

```
<p>Prevent prevent here box speak level.<br/>Like send serve remember beyond newspaper effort debate.</p><p>Instead choose option contain air less various.<br/>Happen body nearly kitchen growth purpose those like.</p><p><br/>Involve still along stock series spend.</p>
```

This approach is fairly naive, but usable if you're not too worried about edge cases.
